GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using GitCommands;
+using Microsoft.Win32;
+
+namespace GitUI.CommandsDialogs.SettingsDialog.ShellExtension
+{
+    public static class ShellExtensionManager
+    {
+        /// <summary>
+        /// Checks if necessary for registration files were found
+        /// </summary>
+        /// <returns>false only if all necessary files were found, otherwise true</returns>
+        public static bool IsNotInstalled()
+        {
+            string path32 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx32Name);
+            string path64 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx64Name);
+            return string.IsNullOrEmpty(path32) || (Environment.Is64BitOperatingSystem && string.IsNullOrEmpty(path64));
+        }
+
+        /// <summary>
+        /// Checks if shell extension is properly registered in system
+        /// </summary>
+        /// <returns>false only if all checks are positive, otherwise true</returns>
+        public static bool IsNotRegistered()
+        {
+            return string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.LocalMachine, @"Software\Microsoft\Windows\CurrentVersion\Shell Extensions\Approved",
+                    "{3C16B20A-BA16-4156-916F-0A375ECFFE24}")) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"*\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\shellex\ContextMenuHandlers\GitExtensions2", null)) ||
+                string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot,
+                    @"Directory\Background\shellex\ContextMenuHandlers\GitExtensions2",
+                    null));
+        }
+
+        /// <summary>
+        /// Register shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog</exception>
+        public static void RegisterShellExtension() => RegisterAllShellExtensionDlls("{0}");
+
+        /// <summary>
+        /// Register shell extensions.
+        /// The same as <see cref="RegisterShellExtension"/> but throws no exceptions
+        /// </summary>
+        public static void RegisterShellExtensionSafe()
```

Comment:
In .NET lingo this should be look something like the following returning a flag indicating success or failure.
```suggestion
        public static bool TryRegister()
```
Though at this stage I'm not sure I understand the need for this.

Reply:
I don't like to introduce return value just in case someone need it or because of naming. I didn't need return value so it's void. That's why it _Safe_ (in meaning not exceptions will be thrown as described in xml doc).
But if you think it's better to introduce not used return value I'll fix it.

Reply:
P.S. I decided that it's better to get rid of *Safe methods.